pkg/agent/secondarynetwork/podwatch/controller.go

```diff
@@ -114,6 +124,53 @@ func podKeyGet(pod *corev1.Pod) string {
 	return pod.Namespace + "/" + pod.Name
 }
 
+// buildVFDeviceIDListPerPod is a helper function to build a cache structure with the
+// list of all the PCI addresses allocated per Pod based on their resource requests (in Pod spec).
+// When there is a request for a VF resource (to associate it for a secondary network interface),
+// getUnusedSriovVFDeviceIDPerPod will use this cache information to pick up a unique PCI address
+// which is still not associated with a network device name.
+// NOTE: buildVFDeviceIDListPerPod is called only if a Pod specific VF to Interface mapping cache
+// was not build earlier. Sample initial entry per Pod: "{18:01.1,""},{18:01.2,""},{18:01.3,""}"
+func (pc *PodController) buildVFDeviceIDListPerPod(podName, podNamespace string) error {
+	key := podNamespace + "/" + podName
+	pc.vfDeviceIDCacheMutex.RLock()
+	defer pc.vfDeviceIDCacheMutex.RUnlock()
+	if pc.vfDeviceIDUsageMap[key] != nil {
+		return nil
+	}
+	podSriovVFDeviceIDs, err := getPodContainerDeviceIDs(podName, podNamespace)
+	if err != nil {
+		return fmt.Errorf("getPodContainerDeviceIDs failed: %v", err)
+	}
+	var vfDeviceIDInfo []podSriovVFDeviceIDInfo
+	for _, pciAddress := range podSriovVFDeviceIDs {
+		initSriovVfDeviceID := podSriovVFDeviceIDInfo{vfDeviceID: pciAddress, ifName: ""}
+		vfDeviceIDInfo = append(vfDeviceIDInfo, initSriovVfDeviceID)
+		pc.vfDeviceIDUsageMap[key] = vfDeviceIDInfo
```

Comment:
Why update the map in each iteration? We just need to update it once after the loop?

Reply:
And we need write lock for the update.

Reply:
right. update on each iteration is not required. modified with along with the write lock.